cdap-securestore-spi/src/main/java/co/cask/cdap/securestore/spi/secret/SecretMetadata.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.securestore.spi.secret;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Represents metadata for the sensitive data to be stored.
+ */
+public class SecretMetadata {
+  private final String name;
+  private final String description;
+  private final long creationTimeMs;
+  private final Map<String, String> properties;
+
+  /**
+   * Constructs metadata with provided secret name, description, creation time and properties.
+   *
+   * @param name the name of the secret to which this metadata belongs to
+   * @param description description of the secret to which this metadata belongs to
+   * @param creationTimeMs creation time of the secret in milli seconds
+   * @param properties properties of the secret
+   */
+  public SecretMetadata(String name, String description, long creationTimeMs, Map<String, String> properties) {
+    this.name = name;
+    this.description = description;
+    this.creationTimeMs = creationTimeMs;
+    this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+  }
+
+  /**
+   * @return name of the secret
+   */
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * @return description of the secret
+   */
+  public String getDescription() {
+    return description;
+  }
+
+  /**
+   * @return creation time of the secret in milli seconds
+   */
+  public long getCreationTimeMs() {
+    return creationTimeMs;
+  }
+
+  /**
+   * @return properties of the secret
+   */
+  public Map<String, String> getProperties() {
+    return properties;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SecretMetadata that = (SecretMetadata) o;
+    return creationTimeMs == that.creationTimeMs &&
+      Objects.equals(name, that.name) &&
+      Objects.equals(description, that.description) &&
+      Objects.equals(properties, that.properties);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(name, description, creationTimeMs, properties);
+  }
+
+  @Override
+  public String toString() {
+    return "SecretMetadata{" +
+      "name='" + name + '\'' +
+      ", description='" + description + '\'' +
+      ", creationTimeMs=" + creationTimeMs +
+      '}';
```

Comment:
This does not print properties?

Reply:
no because properties can include specific details like crypto key version being used for encryption

Reply:
If that is a sensitive information then should not be stored as a property in itself and be a secret. Not printing it does not make it secure I think.

Reply:
The properties are for added information about the sensitive data. Properties like crypto key name is not sensitive data. However, toString can be used for logging purpose, thats why properties are omitted from toString.

If implementation want to add it, toString should be overridden.

Reply:
Will the SPI implementation override the toString of SecretMetadata? That does not seem correct.

Reply:
Why not?

Reply:
Then an implementation can override Secret class too?

Ideally this class and Secret class should be final.

Reply:
`SecretMetadata` is just a data holder class right? As far as spi contract is being honored, why make class final?

Reply:
> SecretMetadata is just a data holder class right? As far as spi contract is being honored, I wonder if it would matter if class had one method overridden?

I don't think there is a need for this class to be extensible and neither it is specifically designed for it. Hence it is better to make it final.
For details please refer to 
Effective Java Item 17: Design and document for inheritance or else prohibit it. 

> I do have strong opinion on stringifying properties, but the reason for not adding it is to be more conservative.

Unless there is a strong need (as in Secret) for having toString print something else I will prefer it to have give an actual representation of the class rather than hide the info. For some user who might end up storing sensitive information we are hiding and confusing all the other remaining users.

Reply:
I took reference from https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/java/security/KeyStore.java#L673 - which does not print attributes of an entry. I think not printing them is safest.

Reply:
That class toString does not seem to print the object at all so should not be taken as reference. Due to the lack of documentation on that method hard to say what is the intention there.

If it is about being safe because crypto key version can be stored in property then someone can even store it in description so should we not print it too?

Reply:
It does print Algorithm of the provided key but not attributes. 

Not dragging this discussion further, we can revisit this case when other implementations (besides default implementation) requires properties in toString. For now, we can keep it as is.

Reply:
@CuriousVini - let's resolve the comments before merging and if it needs offline discussions to move things faster let's do that. In this case +1 on not printing the properties.